pareto/utilities/cm_utils/opt_utils.py

```diff
@@ -0,0 +1,196 @@
+# publicly and display publicly, and to permit others to do so.
+#####################################################################################################
+"""
+Functions used in run_infrastructure_analysis.py
+"""
+
+from pareto.other_models.CM_module.models.qcp_br import build_qcp_br
+import pyomo.environ as pyo
+from pareto.utilities.cm_utils.gen_utils import obj_fix
+
+
+def max_theoretical_recovery_flow_opt(
+    model, treatment_unit, desired_li_conc, tee=False
+):
+    """This function computes the largest flow possible to
+    the treatment unit while still keeping the Li
+    concentration above the desired level.
+
+    This function ignores all infrastructure - this is only
+    based on mass balance along.
+
+    This function differs from max_theoretical_recovery_flow
+    since this function uses an LP and Gurobi to find the value
+    """
+    assert model.p_alpha[treatment_unit, "Li"] > 0.999
+    alphaW = model.p_alphaW[treatment_unit]
+    # desired concentration at the inlet - this will be corrected at the end
+    desired_li_conc = desired_li_conc * (1 - alphaW)
+
+    mm = pyo.ConcreteModel()
+    mm.S = list(model.p_FGen.index_set())
+    bounds = {k: (0, model.p_FGen[k] * 7) for k in mm.S}
+    mm.F = pyo.Var(mm.S, bounds=bounds)
+    mm.cumulative_F = pyo.Var(bounds=(10, None))
+    mm.cumulative_F_con = pyo.Constraint(
+        expr=mm.cumulative_F == sum(mm.F[t] for t in mm.S)
+    )
+    mm.obj = pyo.Objective(expr=mm.cumulative_F, sense=pyo.maximize)
+    mm.total_li = pyo.Expression(
+        expr=sum(mm.F[t] * model.p_CGen[t[0], "Li", t[1]] for t in mm.S)
+    )
+    mm.quality_con = pyo.Constraint(
+        expr=mm.total_li >= desired_li_conc * mm.cumulative_F
+    )
+    # mm.overall_conc = pyo.Expression(expr = mm.total_li/sum(mm.F[t] for t in mm.S))
+
+    status = pyo.SolverFactory("ipopt").solve(mm, tee=tee)
+    pyo.assert_optimal_termination(status)
+    return pyo.value(mm.cumulative_F) * (1 - alphaW)
+
+
+def max_recovery_with_infrastructure(data, tee=False):
+    # build the model from the loaded data
+    model = build_qcp_br(data)
+
+    ###
+    # First, we solve a flow-based LP without any concentration variables
+    ###
+    model.obj.deactivate()
+    model.br_obj.deactivate()
+    model.treatment_only_obj.activate()
+
+    # create lists of constraints involving concentrations and all flow / inventory variables
+    conclist = [
+        model.MSconc,
+        model.Pconc,
+        model.Cconc,
+        model.Sconc,
+        model.Wconc,
+        model.TINconc,
+        model.NTTWconc,
+        model.NTCWconc,
+        model.NTSrcconc,
+        model.minconccon,
+    ]
+    flowlist = [
+        model.Cflow,
+        model.noinvflow,
+        model.Sinv,
+        model.Dflow,
+        model.Wflow,
+        model.TINflow,
+        model.NTTWflow,
+        model.NTCWflow,
+    ]
+
+    # I think this fixes the concentration variables and removes the concentration constraints
```

Comment:
You think or you know? Please avoid expressing this kind of uncertainty in code comments - it will only confuse future developers  (and possibly users)looking at the code.